### PR TITLE
Clean up Task parameter validation logic

### PR DESCRIPTION
**Pull request by lbernick**
This commit adds docstrings for many functions in the substitution package and in task_validation.go. It also cleans up the substitution package by removing unused functions, adding tests for functions that lack them, renaming functions to better reflect their purpose, and replacing calls to regexp.MustCompile (which can panic) with regexp.Compile (which does not panic). It cleans up task_validation.go by removing functions that are just wrappers around functions in the substitution package, and splitting ValidateParameterVariables into several helpers. No functional changes expected as a result of this commit.

This refactoring will help to address https://github.com/tektoncd/pipeline/issues/6647.
/kind cleanup

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- n/a Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) if any changes are user facing, including updates to minimum requirements e.g. Kubernetes version bumps
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings). See some examples of [good release notes](https://github.com/tektoncd/community/blob/main/standards.md#good-release-notes).
- n/a Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
Some functions in pkg/substitution have been removed or renamed.
```

### CODE REVIEW

File: pkg/apis/pipeline/v1/task_validation.go [hunk 1]
```diff
@@ -362,39 +362,71 @@ func (p ParamSpec) ValidateObjectType(ctx context.Context) *apis.FieldError {
...
-// ValidateParameterVariables validates all variables within a slice of ParamSpecs against a slice of Steps
-func ValidateParameterVariables(ctx context.Context, steps []Step, params []ParamSpec) *apis.FieldError {
-	allParameterNames := sets.NewString()
-	stringParameterNames := sets.NewString()
-	arrayParameterNames := sets.NewString()
-	objectParamSpecs := []ParamSpec{}
-	var errs *apis.FieldError
+// sortParamsByType splits the input params into string params, array params, and object params, in that order
+func sortParamsByType(params []ParamSpec) ([]ParamSpec, []ParamSpec, []ParamSpec) {
+	var stringParams []ParamSpec
```
Comment (dibyom) [inline pkg/apis/pipeline/v1/task_validation.go:367]: nit: 
```suggestion
	var stringParams, arrayParams, objectParams []ParamSpec
```

**Review by dibyom — APPROVED** [review]

**Comment by dibyom** [comment]
> message: 'The step "params-applied" in TaskRun "step-script-bdzk8" failed to pull
              the image "". The pod errored with the message: "Back-off pulling image "python"."'

Test failure is due to network connectivity issue

File: pkg/apis/pipeline/v1/task_validation.go [hunk 2]
```diff
@@ -362,39 +362,69 @@ func (p ParamSpec) ValidateObjectType(ctx context.Context) *apis.FieldError {
...
+	return stringParams, arrayParams, objectParams
+}
+
+func getNames(params []ParamSpec) []string {
+	var names []string
+	for _, p := range params {
+		names = append(names, p.Name)
+	}
+	return names
+}
```
Comment (jerop) [inline pkg/apis/pipeline/v1/task_validation.go:389]: can this be a member function of `ParamSpecs`? I'd imagine that this is useful elsewhere, and most likely this code  is duplicated

File: pkg/apis/pipeline/v1/task_validation.go [hunk 3]
```diff
@@ -362,39 +362,69 @@ func (p ParamSpec) ValidateObjectType(ctx context.Context) *apis.FieldError {
...
+
+// validateNoDuplicateParamNames returns an error if any of the params have the same name
+func validateNoDuplicateParamNames(params []ParamSpec) *apis.FieldError {
```
Comment (jerop) [inline pkg/apis/pipeline/v1/task_validation.go:392]: can this also be a member function of `ParamSpecs`?

File: pkg/apis/pipeline/v1/task_validation.go (hunk 2 above)
Comment (lbernick) [reply pkg/apis/pipeline/v1/task_validation.go]: done

File: pkg/apis/pipeline/v1/task_validation.go (hunk 3 above)
Comment (lbernick) [reply pkg/apis/pipeline/v1/task_validation.go]: done

File: pkg/substitution/substitution_test.go [hunk 4]
```diff
@@ -340,6 +319,90 @@ func TestValidateEntireVariableProhibitedP(t *testing.T) {
 	}
 }
 
+func TestValidateVariableIsolatedP(t *testing.T) {
```
Comment (jerop) [inline pkg/substitution/substitution_test.go:322]: could also update the test names please?

Comment (lbernick) [reply pkg/substitution/substitution_test.go]: done

**Review by chitrangpatel — APPROVED** [review]
Thanks for this Lee! It makes it much more readable.

**Pull request merged**